accepted/00000-http-api.md

```diff
@@ -0,0 +1,214 @@
+
+```java
+// GET https://uyuni.server/rhn/manager/api/activationkey/get_details?key=1-DEFAULT
+
+get("/activationkey/get_details", (req, res) -> {
+  // 1.Unwrap the session user, and the querystring parameter named "key"
+  // 2.Call to the `getDetails` method, passing the unwrapped arguments
+  // 3.Serialize and return the JSON result
+});
+```
+
+### Example 2
+
+Consider the following XMLRPC handler method:
+```java
+public ContentProject updateProject(User user, String label, Map<String, Object> props)
+```
+
+With the proposed annotations, it may be written as:
+```java
+@XmlRpcApiEndpoint
+@HttpApiEndpoint
+public ContentProject updateProject(User user, @RouteParam String label, Map<String, Object> props)
+```
+
+At registration, Spark route wrapper will create the following route definition for this method:
+
+```java
+// POST https://uyuni.server/rhn/manager/api/contentmanagement/update_project/myproject
+
+post("/contentmanagement/update_project/:label", (req, res) -> {
+  // 1.Unwrap the session user, the route parameter named "label",
+  //   and the JSON object named "props" from the request body
+  // 2.Call to the `updateProject` method, passing the unwrapped arguments
+  // 3.Serialize and return the JSON result
+});
+```
+
+
+## 2. Authentication
+
+The web UI performs token-based authentication via a session cookie called `pxt-session-cookie`. Currently, the web UI login page exposes an API endpoint to support this process. This endpoint shall also be used for API authentication.
+
+### Login handshake demo
+
+```bash
+$ curl --include --header "Content-Type: application/json" \
+--data "{'login':'admin','password':'admin'}" \
+https://uyuni.dallas/rhn/manager/api/login
+
+HTTP/1.1 200 200
+...
+Set-Cookie: pxt-session-cookie=209x9fa00e2e947ace87de0743f6cf63b8b0b4bfbfca2310ce6829941fa9db947987; Max-Age=3600; Expires=Tue, 08-Feb-2022 20:05:41 GMT; Path=/; Secure; HttpOnly;HttpOnly;Secure
+...
+
+$ curl --include --cookie "pxt-session-cookie=209x9fa0..." \
+https://uyuni.server/rhn/manager/api/systems/keys
+
+HTTP/1.1 200 200
+...
+<JSON data follows>
+```
+
+
+## 3. I/O
+
+On the HTTP API side, I/O serialization is mostly handled by Spark, and the existing `GSON` library. XMLRPC uses specialized "serializer" classes to write structured output on XML. These serializer classes shall be used exclusively for the XMLRPC output.
+
+
+## 4. Exception Handling
+
+The current implementation of the business logic uses exceptions that are designed specifically for XMLRPC (see: `FaultException` and subclasses). To be able to use the business logic with both interfaces, a new, internal exception structure shall be designed to offer enough flexibility in exception handling. These exceptions are then translated to the specific interface, and processed accordingly. For HTTP endpoints, the internal exceptions can be represented by different HTTP status codes where appropriate.
```

Comment:
No specific comment here, just a note that would be good to keep in mind if we can standardize/centralize this in the current network layer on the frontend. Currently, error and error message handling is a bit of a bazaar where everything is mix and match.